[RDY] Always enable logging

justinmk opened the pull request:
Goals of this PR:

- [x] change the default log location to `~/.local/share/nvim/log`
- [x] set NVIM_LOG_FILE to the default path if it is undefined
- [x] Establish ERROR log level as critical, rare errors that must be logged. There are cases where we have no visibility in hangs/crashes because logging is disabled. 
    - With this established, log file for end-users should be small or non-existent, unless there was a severe problem.
- [x] Always enable logging, even for "Release" builds. If logging is not noisy then this doesn't affect end-users.
- [x] remove DISABLE_LOG
- [x] Travis CI: dump the contents of the log file after each run
- [x] documentation

Future PR:

- `nvim_log(level, message)` API function
- update API clients (python, etc.) to use `nvim_log()` for critical messages
- `stdpath('log')` ?
- log Vim errors that happen during shutdown https://github.com/neovim/neovim/pull/6827#issuecomment-305603694

justinmk committed:
log: log_uv_handles

justinmk committed:
loop_close: Avoid infinite loop, and log it.

Avoids a hang, and also helps diagnose issues like:

https://github.com/neovim/neovim/pull/6594#issuecomment-298321826

justinmk committed:
log: Always enable; remove DISABLE_LOG

- Establish ERROR log level as "critical". Such errors are rare and will
  be valuable when users encounter unusual circumstances.
- Set -DMIN_LOG_LEVEL=3 for release-type builds

justinmk committed:
ci: Dump $NVIM_LOG_FILE contents

justinmk commented on src/nvim/event/loop.c:
```diff
@@ -85,9 +86,18 @@ void loop_close(Loop *loop, bool wait)
+      // Some libuv resource was not correctly deref'd. Log and bail.
+      log_uv_handles(&loop->uv);
+      break;
+    }
```
If loop_close takes longer than 2 seconds it is most likely an infinite loop. So, log the state and exit the loop.

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -370,7 +370,7 @@ static void receive_msgpack(Stream *stream, RBuffer *rbuf, size_t c,
     snprintf(buf, sizeof(buf), "ch %" PRIu64 " was closed by the client",
              channel->id);
-    call_set_error(channel, buf);
+    call_set_error(channel, buf, WARNING_LOG_LEVEL);
```
This is  usually not a failure mode, but a normal situation. So don't log it at ERROR level.

oni-link commented on src/nvim/garray.h:
```diff
@@ -37,7 +37,7 @@ typedef struct growarray {
 {
   if ((int)item_size != gap->ga_itemsize) {
-    ELOG("wrong item size in garray(%d), should be %d", item_size);
+    WLOG("wrong item size in garray(%d), should be %d", item_size);
```
Are we missing a size argument here?

justinmk replied:
Looks like it :)

justinmk committed:
ga_append_via_ptr: fix log arguments

justinmk committed:
log: set $NVIM_LOG_FILE; fallback to $XDG_DATA_HOME/nvim/log

justinmk committed:
test: Set $NVIM_LOG_FILE to test-local path

- Do not delete it: may need to inspect it after tests finished.
- Avoids writing to stderr in cases where the test-local $XDG_DATA_HOME
  was not created yet.

justinmk committed:
log: Fall back to CWD-relative .nvimlog

If if the resolved $NVIM_LOG_FILE *and* stdpath("data")/log cannot be
created (e.g. because the XDG data directory does not exist), fall back
to .nvimlog in the current direcrtory.

justinmk committed:
test: iswin(): detect without nvim session

blueyed commented:
Nice!
btw: would it be possible to also log Vim errors, especially when they happen on shutdown/quitting?
(I've just seen some when quitting Neovim slightly pop up in red, but then it is gone already)

blueyed commented on .travis.yml:
```diff
@@ -21,8 +21,6 @@ env:
     # Log directory for Clang sanitizers and Valgrind.
     - LOG_DIR="$BUILD_DIR/log"
-    # Nvim log file.
-    - NVIM_LOG_FILE="$BUILD_DIR/.nvimlog"
```
Seems to be cleaner to keep it here, no?
Is `$BUILD_DIR` already setup here (defined above)?  (but it seems to work for `$LOG_DIR`?!

justinmk replied:
@blueyed That's part of the temporary commits, I am puzzled about why travis is failing.

oni-link commented on src/nvim/os_unix.c:
```diff
@@ -141,7 +141,8 @@ void mch_exit(int r) FUNC_ATTR_NORETURN
 
-  event_teardown();
+  // Exit with error if main_loop did not teardown gracefully.
+  r = event_teardown() ? r : (r || 1);
```
Why use `(r||1)` here, it is always `1`?

justinmk replied:
@oni-link You're right, that needs to be a ternary. The intention is to preserve the value of `r > 0`.

oni-link commented on src/nvim/log.c:
```diff
@@ -7,52 +7,74 @@
+      || os_isdir((char_u *)log_file_path)
+      || !log_try_create(log_file_path)) {
+    // Invalid $NVIM_LOG_FILE or failed to expand; fall back to default.
+    memset(log_file_path, 0, size);
```
This could be removed. The following `xstrlcpy()` will overwrite this buffer and terminate it with `NUL`.

oni-link commented on src/nvim/log.c:
```diff
@@ -7,52 +7,74 @@
+    xfree(defaultpath);
+    // Fall back to .nvimlog
+    if (len >= size || !log_try_create(log_file_path)) {
+      len = xstrlcpy(log_file_path, ".nvimlog", sizeof(".nvimlog"));
```
`sizeof(".nvimlog")` is not the size that is expected here, it should be the size of the destination buffer.

oni-link commented on src/nvim/log.c:
```diff
@@ -7,52 +7,74 @@
+    }
+    // Fall back to stderr
+    if (len >= size || !log_try_create(log_file_path)) {
+      memset(log_file_path, 0, size);
```
`log_file_path[0]=NUL;` should be enough to mark the buffer as empty.

justinmk committed:
doc: *standard-path*, *$NVIM_LOG_FILE*

jdebp commented:
`Development-tips` on the wiki needs to be brought up-to-date.
